docs/man/man3/CanCom.3
.TH "CanCom" 3 "Tue May 17 2022" "EPS MCU" \" -*- nroff -*-
.ad l
.nh
.SH NAME
CanCom \- Modtagning og afsendelse af data over CAN\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <CanCom\&.h>\fP
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBCanMessageId\fP { \fBCAN_TIME_SYNC\fP = 1, \fBCAN_SET_CONFIG\fP, \fBCAN_GET_CONFIG\fP, \fBCAN_GET_READING\fP, \fBCAN_LOG\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBCanCom\fP (\fBNoMutexCAN\fP *\fBcan\fP)"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static void \fBSendBuffer\fP (size_t size)"
.br
.RI "Sender fra serializeret buffer over CAN ved at dele hvert frame op i 8 bytes\&. "
.ti -1c
.RI "static void \fBSendReadings\fP ()"
.br
.RI "Send alle målinger i CircularBuffer'eren (adc_buffer_) over CAN\&. "
.ti -1c
.RI "static \fBAdcRead::adc_reading\fP \fBGetReading\fP ()"
.br
.RI "Modtag en enkelt ADC måling fra serialized_buffer_\&. "
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static uint8_t \fBserialized_buffer_\fP [\fBsend_reading_size\fP *\fBADC_BUFFER_SIZE\fP]"
.br
.ti -1c
.RI "static \fBNoMutexCAN\fP * \fBcan_\fP"
.br
.in -1c
.SS "Static Private Member Functions"

.in +1c
.ti -1c
.RI "static size_t \fBGetFrameSize\fP ()"
.br
.RI "Find størrelsen af en modtaget Protobuf pakke inden decode\&. "
.ti -1c
.RI "static void \fBShiftBuffer\fP (size_t shift_bytes)"
.br
.RI "Rykker alle bytes i serialized_buffer_ med et antal pladser\&. "
.in -1c
.SH "Detailed Description"
.PP 
Modtagning og afsendelse af data over CAN\&. 


.PP
\fBNote\fP
.RS 4
Hånterer ikke rent faktisk modtagelse af CAN da dette gøres i \fBmain\&.cpp\fP 
.RE
.PP

.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBCanCom::CanMessageId\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fICAN_TIME_SYNC \fP\fP
.TP
\fB\fICAN_SET_CONFIG \fP\fP
.TP
\fB\fICAN_GET_CONFIG \fP\fP
.TP
\fB\fICAN_GET_READING \fP\fP
.TP
\fB\fICAN_LOG \fP\fP
.SH "Constructor & Destructor Documentation"
.PP 
.SS "CanCom::CanCom (\fBNoMutexCAN\fP * can)\fC [inline]\fP"

.SH "Member Function Documentation"
.PP 
.SS "size_t CanCom::GetFrameSize ()\fC [static]\fP, \fC [private]\fP"

.PP
Find størrelsen af en modtaget Protobuf pakke inden decode\&. 
.PP
\fBReturns\fP
.RS 4
size_t størrelsen af pakken i bytes
.RE
.PP
\fBNote\fP
.RS 4
Anvdendes ikke da vi ikke modtager pakker der fylder mere end 1 CAN frame 
.RE
.PP

.SS "\fBAdcRead::adc_reading\fP CanCom::GetReading ()\fC [static]\fP"

.PP
Modtag en enkelt ADC måling fra serialized_buffer_\&. 
.PP
\fBReturns\fP
.RS 4
\fBAdcRead::adc_reading\fP decoded adc reading
.RE
.PP
\fBNote\fP
.RS 4
Anvdendes ikke da vi ikke modtager ADC målinger igennem CAN 
.RE
.PP

.SS "void CanCom::SendBuffer (size_t size)\fC [static]\fP"

.PP
Sender fra serializeret buffer over CAN ved at dele hvert frame op i 8 bytes\&. 
.PP
\fBParameters\fP
.RS 4
\fIsize\fP Antale bytes at sende fra bufferen 
.RE
.PP

.SS "void CanCom::SendReadings ()\fC [static]\fP"

.PP
Send alle målinger i CircularBuffer'eren (adc_buffer_) over CAN\&. Hver måling i adc_buffer_ enkodes med Google Protobuf structen \fIsend_reading og serialiseres\&. De serialiserede bytes lægges ind i serialized_buffer\fP\&. Hvis der ikke blev målt på mppt og rails i en given måling sættes has_mppt_x_x og has_sub_x_x bools'ne til False, og protobufs sørger for at serialiseringen fylder mindre\&. 
.SS "void CanCom::ShiftBuffer (size_t shift_bytes)\fC [static]\fP, \fC [private]\fP"

.PP
Rykker alle bytes i serialized_buffer_ med et antal pladser\&. 
.PP
\fBParameters\fP
.RS 4
\fIshift_bytes\fP antal pladser at rykke pladser i bufferen 
.RE
.PP

.SH "Member Data Documentation"
.PP 
.SS "\fBNoMutexCAN\fP * CanCom::can_\fC [static]\fP"

.SS "uint8_t CanCom::serialized_buffer_\fC [static]\fP"


.SH "Author"
.PP 
Generated automatically by Doxygen for EPS MCU from the source code\&.


lib/CanCom/CanCom.cpp
//
// Created by EIT-416 on 5/3/22.
//

#include "CanCom.h"
#include "AdcRead.h"
#include "mbed.h"
#include <pb_encode.h>
#include <pb_decode.h>
#include "reading.pb.h"
#include "ProjectConfig.h"
#include "Log.h"

uint8_t CanCom::serialized_buffer_[send_reading_size * ADC_BUFFER_SIZE];

NoMutexCAN *CanCom::can_;

/**
 * @brief Send alle målinger i CircularBuffer'eren (adc_buffer_) over CAN.
 * 
 * Hver måling i adc_buffer_ enkodes med Google Protobuf structen _send_reading
 * og serialiseres. De serialiserede bytes lægges ind i serialized_buffer_. Hvis der
 * ikke blev målt på mppt og rails i en given måling sættes has_mppt_x_x og 
 * has_sub_x_x bools'ne til False, og protobufs sørger for at serialiseringen
 * fylder mindre.
 * 
 */
void CanCom::SendReadings() {
  auto size = AdcRead::adc_buffer_.size();


  AdcRead::adc_reading buf{};
  send_reading to_send;
  pb_ostream_t encode = pb_ostream_from_buffer(\
  serialized_buffer_, sizeof(serialized_buffer_));

  for (size_t i = 0; i < size; ++i) {
    AdcRead::adc_buffer_.pop(buf);
    to_send.bat_v_1 = buf.bat_v_1;
    to_send.bat_i_1 = buf.bat_i_1;
    to_send.timestamp = buf.timestamp;

    if (!buf.mppt_v_1) { // Don't send data if it wasn't read
      to_send.has_mppt_v_1 = false;
      to_send.has_mppt_i_1 = false;
      to_send.has_sub_v_1 = false;
      to_send.has_sub_i_1 = false;
    } else {
      to_send.has_mppt_v_1 = true;
      to_send.has_mppt_i_1 = true;
      to_send.has_sub_v_1 = true;
      to_send.has_sub_i_1 = true;

      to_send.mppt_v_1 = buf.mppt_v_1;
      to_send.mppt_i_1 = buf.mppt_i_1;
      to_send.sub_v_1 = buf.sub_v_1;
      to_send.sub_i_1 = buf.sub_i_1;
    }

    pb_encode(&encode, send_reading_fields, &to_send);

  }

  SendBuffer(encode.bytes_written);

}

/**
 * @brief Sender fra serializeret buffer over CAN ved at dele hvert
 *        frame op i 8 bytes
 * 
 * @param size Antale bytes at sende fra bufferen
 */
void CanCom::SendBuffer(size_t size) {
  auto bytes = size;
  auto can_frames = (bytes / 8) + ((bytes % 8) ? 1 : 0);
  uint8_t framebuffer[8];

  //  CANMessage *msg;
  LOG(LOG_DEBUG, "Sending CAN");
  for (size_t i = 0; i < can_frames; ++i) {
    auto frame_size = (bytes / 8) ? 8 : bytes;
    for (size_t j = 0; j < frame_size; ++j) {
      framebuffer[j] = serialized_buffer_[j];
    }
    ShiftBuffer(frame_size);
    CANMessage msg(CAN_GET_READING, framebuffer, frame_size);
    can_->write(msg);
    bytes -= frame_size;
  }
  LOG(LOG_DEBUG, "Done Sending");

}

/**
 * @brief Find størrelsen af en modtaget Protobuf pakke inden decode
 * 
 * @return size_t størrelsen af pakken i bytes
 * 
 * @note Anvdendes ikke da vi ikke modtager pakker der fylder mere end 1 CAN frame
 */
size_t CanCom::GetFrameSize(){
  if (serialized_buffer_[0] != '\b') {
    LOG(LOG_ERROR,"Can read buffer did not start with STX");
    return 0; // TODO better Error Reporting
  }

  for (int i = 1; i < send_reading_size; ++i) {
    auto index = serialized_buffer_[i];
    if (index == '\b' or index == '\0') {
      return i ;
    }
  }
}

/**
 * @brief Modtag en enkelt ADC måling fra serialized_buffer_
 * 
 * @return AdcRead::adc_reading decoded adc reading
 * 
 * @note Anvdendes ikke da vi ikke modtager ADC målinger igennem CAN
 */
AdcRead::adc_reading CanCom::GetReading() {
  AdcRead::adc_reading ret{};
  send_reading buf;

  auto frame_size = GetFrameSize();
  pb_istream_t decode = pb_istream_from_buffer(\
  serialized_buffer_, frame_size);


  pb_decode(&decode, send_reading_fields, &buf);

  ret.bat_v_1 = buf.bat_v_1;
  ret.bat_i_1 = buf.bat_i_1;
  ret.timestamp = buf.timestamp;

  if (buf.has_mppt_v_1) { // Don't send data if it wasn't read
    ret.mppt_v_1 = buf.mppt_v_1;
    ret.mppt_i_1 = buf.mppt_i_1;
    ret.sub_v_1 = buf.sub_v_1;
    ret.sub_i_1 = buf.sub_i_1;
  }

//  Shift buffer by amount read
  ShiftBuffer(frame_size);
  return ret;
}

/**
 * @brief Rykker alle bytes i serialized_buffer_ med et antal pladser
 * 
 * @param shift_bytes antal pladser at rykke pladser i bufferen
 */
void CanCom::ShiftBuffer(size_t shift_bytes) {
  for (auto i = shift_bytes; i < sizeof(serialized_buffer_) - shift_bytes; ++i) {
    serialized_buffer_[i - shift_bytes] = serialized_buffer_[i];
    serialized_buffer_[i] = '\0';
  }
}
